metadata/src/main/java/org/apache/kafka/controller/ControllerRequestContext.java

```diff
@@ -42,19 +45,39 @@ public static OptionalLong requestTimeoutMsToDeadlineNs(
     private final OptionalLong deadlineNs;
     private final RequestHeaderData requestHeader;
 
+    private final Optional<Consumer<Double>> requestedPartitionCountRecorder;
+
```

Comment:
We add this here because `class ControllerMutationQuotaManager` and `trait ControllerMutationQuota` are written in Scala while the KRaft controller and such are written in Java with no access to the Scala class/trait -- we can't pass in anything without wrapping it in Java somehow.  All we need to do is record a request for some number of partitions, and this matches that requirement precisely.

Reply:
We could have an interface in server-common that the scala classes implement.

Reply:
Yes, this is messy. I have a lot of ideas for improving this, but most of them would require substantial refactoring.

For now, let's just get rid of the obvious bad stuff.
* Using a double for number of partitions is a bit silly
* "Recorder" is also a bit of a misnomer -- we are doing more than recording the number of partitions.
* The Optional isn't needed -- just pass a Consumer that does nothing if you don't want to do anything.

So maybe change it to something like `Consumer<Integer> applyPartitionChangeQuota` ?